src/java/org/apache/cassandra/service/accord/serializers/TableMetadatasAndKeys.java

```diff
@@ -0,0 +1,270 @@
+
+public class TableMetadatasAndKeys extends IVersionedWithKeysSerializer.AbstractWithKeysSerializer implements Sliceable<TableMetadatasAndKeys>
+{
+    public static class KeyCollector extends AbstractSortedCollector<PartitionKey, Keys>
+    {
+        private static final Comparator<Object> comparator = Comparator.comparing(v -> ((PartitionKey) v));
+
+        public final TableMetadatas tables;
+
+        public KeyCollector(TableMetadatas tables)
+        {
+            this.tables = tables;
+        }
+
+        public TableMetadatasAndKeys buildTablesAndKeys()
+        {
+            return new TableMetadatasAndKeys(tables, build());
+        }
+
+        @Override
+        Comparator<Object> comparator()
+        {
+            return comparator;
+        }
+
+        public PartitionKey collect(TableMetadata table, DecoratedKey key)
+        {
+            TableId tableId = tables.get(table.id).id;
+            if (count == 1)
+            {
+                PartitionKey one = (PartitionKey) buffer;
+                if (one.prefix() == table && one.partitionKey().equals(key))
+                    return one;
+            }
+            return collect(new PartitionKey(tableId, key));
+        }
+
+        @Override
+        Keys empty()
+        {
+            return Keys.EMPTY;
+        }
+
+        @Override
+        Keys of(PartitionKey one)
+        {
+            return Keys.of(one);
+        }
+
+        @Override
+        Keys copy(Object[] array, int count)
+        {
+            Key[] result = new Key[count];
+            System.arraycopy(array, 0, result, 0, count);
+            return Keys.ofSortedUnique(result);
+        }
+
+        @Override
+        Keys copyBtree(Object[] btree, int count)
+        {
+            return Keys.ofSortedUnique(new BTreeSet<>(btree, comparator()));
+        }
+    }
+
+    private static final TableMetadatasAndKeys NO_KEYS = new TableMetadatasAndKeys(Multi.NONE, Keys.EMPTY);
+    private static final TableMetadatasAndKeys NO_RANGES = new TableMetadatasAndKeys(Multi.NONE, Ranges.EMPTY);
+
+    public static TableMetadatasAndKeys none(Routable.Domain domain)
+    {
+        return domain.isKey() ? NO_KEYS : NO_RANGES;
+    }
+
+    public final TableMetadatas tables;
+    public final Seekables keys;
+
+    public TableMetadatasAndKeys(TableMetadatas tables, Seekables keys)
+    {
+        this.tables = tables;
+        this.keys = keys;
+    }
+
+    public TableMetadata get(TableId tableId)
+    {
+        return tables.get(tableId);
+    }
+
+    public void serializeSeekables(Seekables seekables, DataOutputPlus out) throws IOException
+    {
+        if (keys.domain() == Range)
+        {
+            // we maybe don't guarantee that all exact Range values are included in the superset?? just in case, anyway
+            boolean containsAll = ((AbstractRanges)keys).foldlExact((AbstractRanges) seekables, (r, p, v, i) -> v + 1, 0, 0, 0) == seekables.size();
```

Comment:
so, lets say the following

```
keys = [0, 100], [200, 500]
```

`findNextExactIntersection` is just finding the next intersection, so this `foldlExact` is just a count of intersections?

```
seekables = [0, 10], [15, 20] == 2 matches; so containsAll?
seekables = [0, 10], [300, 400] == 2 matches; so containsAll?
```

Reply:
FoldlExact should use findNextExact which uses compareTo which compares start and end. So it should only contains all if we have an exact match.

Reply:
git shows https://github.com/belliottsmith/cassandra-accord/compare/fc14a154fd514d4ab40b37508fb9497f786835e0...3b24d82e3da5396eb0f3d1be709f85c3f9340690

this calls `routables.foldlexact`, which calls `AbstractRanges::findNextExactIntersection`, which calls `SortedArrays.findNextIntersection`, which calls `accord.primitives.Range#compare`

so, what makes this different than `keys.equals(seekables)`?

Reply:
Subset. It’s for subset serialisation, so we can encode indexes, and this just decides if that’s safe for ranges or not.

Reply:
I guess it's dead code anyway though. Will remove for now.